// vues/lib/webpack/production.config.js
import ExtractTextPlugin       from 'extract-text-webpack-plugin'
import UglifyJSPlugin          from 'uglifyjs-webpack-plugin'
import OptimizeCssAssetsPlugin from 'optimize-css-assets-webpack-plugin'

export default config => {
  return {
    output: {
      filename: '[name].[chunkhash:8].js',
      path: config.buildDir
    },

    module: {
      rules: [
        {
          test: /\.css$/,
          use: ExtractTextPlugin.extract({
            fallback: 'style-loader',
            use: [
              { loader: 'css-loader' },
              {
                loader: 'postcss-loader',
                options: {
                  plugins: () => {
                    return config.postcssOptions
                  }
                }
              }
            ]
          })
        }
      ]
    },

    plugins: [
      new ExtractTextPlugin({
        filename: '[name].[contenthash:8].css'
      }),
      new UglifyJSPlugin({
        sourceMap: false,
        comments: false
      }),
      new OptimizeCssAssetsPlugin({
        cssProcessorOptions: {
          discardComments: { removeAll: true }
        }
      })
    ]
  }
}
